feat(debug): replace DebugElement with new Debug DOM

juliemr opened the pull request:
Now, using `ng.probe(element)` in the browser console returns
a DebugElement when in dev mode.

`ComponentFixture#debugElement` also returns a new DebugElement.

Breaking Change:

This is a breaking change for unit tests. The API for the DebugElement
has changed. Now, there is a DebugElement or DebugNode for every node
in the DOM, not only nodes with an ElementRef. `componentViewChildren` is
removed, and `childNodes` is a list of ElementNodes corresponding to every
child in the DOM. `query` no longer takes a scope parameter, since
the entire rendered DOM is included in the `childNodes`.

Before:

```
componentFixture.debugElement.componentViewChildren[0];
```

After

```
// Depending on the DOM structure of your component, the
// index may have changed or the first component child
// may be a sub-child.
componentFixture.debugElement.children[0];
```

Before:

```
debugElement.query(By.css('div'), Scope.all());
```

After:

```
debugElement.query(By.css('div'));
```

Before:

```
componentFixture.debugElement.elementRef;
```

After:

```
componentFixture.elementRef;
```

juliemr committed:
feat(debug): replace DebugElement with new Debug DOM

Now, using `ng.probe(element)` in the browser console returns
a DebugElement when in dev mode.

`ComponentFixture#debugElement` also returns a new DebugElement.

Breaking Change:

This is a breaking change for unit tests. The API for the DebugElement
has changed. Now, there is a DebugElement or DebugNode for every node
in the DOM, not only nodes with an ElementRef. `componentViewChildren` is
removed, and `childNodes` is a list of ElementNodes corresponding to every
child in the DOM. `query` no longer takes a scope parameter, since
the entire rendered DOM is included in the `childNodes`.

Before:

```
componentFixture.debugElement.componentViewChildren[0];
```

After
```
// Depending on the DOM structure of your component, the
// index may have changed or the first component child
// may be a sub-child.
componentFixture.debugElement.children[0];
```

Before:

```
debugElement.query(By.css('div'), Scope.all());
```

After:

```
debugElement.query(By.css('div'));
```

Before:

```
componentFixture.debugElement.elementRef;
```

After:

```
componentFixture.elementRef;
```

juliemr commented:
@tbosch this is a work in progress, as some tests are not yet upgraded. However, the API changes are complete and the unit tests for the new `DebugElement` are done, and it would be great to get feedback. Particularly, I'm not thrilled with the API for `Renderer#setInjectableMirrors`, which is kind of a grab bag for all sorts of info we need from the view.

cc @jelbourn @wardbell , this is a heads up of some breaking changes to the unit test API. Feedback appreciated.

wardbell commented on modules/angular2/src/core/render/api.ts:
```diff
@@ -42,6 +43,9 @@ export abstract class Renderer implements ParentRenderer {
    */
   abstract setBindingDebugInfo(renderElement: any, propertyName: string, propertyValue: string);
 
+  abstract setInjectableMirrors(renderElement: any, injector: Injector, component: any,
```
This name is deeply mysterious. No doc comments. What does it do?

juliemr replied:
It adds info from the AppElement and AppView (injectable tokens, the injector itself, and the current component) that we'll need for debugging.

It's a renderer function, so it should be internal and a user should never have to bother with it. In any case, it's probably a good idea to add a comment, I'll do that.

tbosch replied:
See my comment at the implementation

juliemr commented:
Remaining tasks

 - [ ] Fix tests that are failing due to the breaking changes
 - [ ] Create before/after examples for the breaking changes in the commit message

juliemr commented:
I've added a commit which fixes Angular's tests, adds a couple necessary features to DebugElement, and removes the ViewListener. Please take a look!

juliemr commented:
I've created a doc to track the changes that will be necessary for the merge to g3: https://docs.google.com/document/d/1wdODXGw8k_bCCYvPzxFxRX7jfAxHn7SXwTCpfsCL07c/

tbosch commented on modules/angular2/src/core/debug/debug_node.ts:
```diff
@@ -0,0 +1,149 @@
+  nativeNode: any;
+  listeners: EventListener[];
+  parent: DebugElement;
+  injectables: any[];
```
Can we change this to be a `directiveTypes: Type[]`?

tbosch replied:
Better: `providerTokens: any[]`

juliemr replied:
done.

tbosch commented on modules/angular2/src/core/debug/debug_node.ts:
```diff
@@ -0,0 +1,149 @@
+      this.childNodes =
+          ListWrapper.concat(ListWrapper.concat(previousChildren, newChildren), nextChildren);
+      for (var i = 0; i < newChildren.length; ++i) {
+        newChildren[i].parent = this;
```
You might also need to remove the children from their previous parents.
And the previous parent could be this element as well.

I.e. first loop over all newChildren and make sure they are removed from their previous parents.
And then add the new ones..

juliemr replied:
done.

tbosch commented on modules/angular2/src/core/debug/debug_node.ts:
```diff
@@ -0,0 +1,149 @@
+  element.childNodes.forEach(node => {
+    if (node instanceof DebugElement) {
+      descendants.push(node);
+      descendants = descendants.concat(_descendants(node));
```
`concat` can get very slow if executed often (we had this once in Dart...). Maybe use `push`?

tbosch replied:
If you pass the predicate to `_descendants` you don't have to concat here at all, only `push` to a result list if the predicate matches

juliemr replied:
done.

tbosch commented on modules/angular2/src/core/debug/debug_node.ts:
```diff
@@ -0,0 +1,149 @@
+  }
+
+  queryAll(predicate: Predicate<DebugElement>): DebugElement[] {
+    var allChildren: DebugElement[] = _descendants(this);
```
Maybe pass the predicate into `_descendants` and filter while walking? Might be a little faster, as we don't have to create an array that contains all nodes.

juliemr replied:
done.

tbosch commented on modules/angular2/src/core/debug/debug_node.ts:
```diff
@@ -0,0 +1,149 @@
+}
+
+// Need to keep the nodes in a global Map so that multiple angular apps are supported.
+export var nativeNodeToDebugNode = new Map<any, DebugNode>();
```
Maybe provide a `getDebugNode(nativeNode: any)` and `getAllDebugNodes()` instead of exporting the `Map`?

juliemr replied:
done.

tbosch commented on modules/angular2/src/core/debug/debug_renderer.ts:
```diff
@@ -0,0 +1,153 @@
+    return this._delegate.setBindingDebugInfo(renderElement, propertyName, propertyValue);
+  }
+
+  setInjectableMirrors(renderElement: any, injector: Injector, component: any, injectables: any[],
```
We need a better way for this!
Maybe `setElementDebugInfo` (Misko agrees)...

And then pass one object (of a new class `RenderDebugInfo`) that contains the injector, component, directiveTypes and locals. Pass that object directly to `DebugElement.setDebugInfo(...)` so you don't need to set the properties of `DebugElement` from outside of `DebugElement`.

juliemr replied:
done.

tbosch commented on modules/angular2/src/core/linker/view.ts:
```diff
@@ -119,6 +119,12 @@ export class AppView implements ChangeDispatcher {
                              (templateName, _) => { localsMap.set(templateName, null); });
     for (var i = 0; i < appElements.length; i++) {
       var appEl = appElements[i];
+      var mirrors = [];
```
Rename to `providerTokens`

juliemr replied:
done.

tbosch commented on modules/angular2/src/platform/dom/debug/ng_probe.ts:
```diff
@@ -0,0 +1,35 @@
+ * Providers which support debugging Angular applications (e.g. via `ng.probe`).
+ */
+export const ELEMENT_PROBE_PROVIDERS: any[] = CONST_EXPR([
+  // DebugNodeStore,
```
Why is this commented out?

juliemr replied:
Forgot to remove it - gone now. Thanks for catching!

tbosch commented on modules/angular2/src/platform/dom/debug/ng_probe.ts:
```diff
@@ -0,0 +1,35 @@
+import {CONST_EXPR, assertionsEnabled, isPresent} from 'angular2/src/facade/lang';
```
This file should now go into `src/platform/debug`, as it is no more platform depending...

tbosch replied:
Also, the `by.ts` should go into `src/platform/debug` now...

juliemr replied:
`by.ts` still uses `DOM` for css matching, so I would keep it in platform/dom/debug. What do you think?

juliemr replied:
We also require `DOM` to set the global variable, and in my current setup I need to get the `DomRootRenderer` from the injector, since otherwise we'd have a `RootRenderer` provider that depended on `RootRenderer`. Is there a better way to create a delegate provider?

tbosch replied:
You are right, let's keep it in the old place. No, I don't know a better way for wrapping a `RootRenderer`, you are right...

tbosch commented on modules/angular2/src/platform/dom/debug/ng_probe.ts:
```diff
@@ -0,0 +1,35 @@
+  return nativeNodeToDebugNode.get(element);
+}
+
+function _createRootRenderer(domRootRenderer) {
```
This can take any `RootRenderer`, not just the `DomRootRenderer`

tbosch commented:
@juliemr this looks already pretty good. Left some comments...

tbosch commented:
Btw, don't worry about the circle CI failure, @IgorMinar said we can ignore it for now...

juliemr commented:
Thanks for the review - all done except for platform location question above.

I squashed the commits and added before/after examples to the breaking changes in the commit message.

juliemr commented:
Ok thanks! This is ready for merge then, with a reminder about the merge doc https://docs.google.com/document/d/1wdODXGw8k_bCCYvPzxFxRX7jfAxHn7SXwTCpfsCL07c/

juliemr commented:
Dear merge-master - this is not quite ready for merge yet. There are some pending g3 issues that I am working on addressing - the doc has been updated.

juliemr commented:
Now ready for merge! I have prepared fixes for all g3 issues, they are linked in the doc.

alexeagle commented:
Missing LGTM - @tbosch ?

tbosch commented on modules/angular2/src/core/debug/debug_renderer.ts:
```diff
@@ -0,0 +1,158 @@
+   * such as <template> placeholders.
+   */
+  setBindingDebugInfo(renderElement: any, propertyName: string, propertyValue: string) {
+    // TODO(juliemr): anything to do here?
```
Remove this todo

alexeagle commented:
merged as e1bf3d3

siddharthjain210 commented:
I have created a sample application with angular2.beta13. Also i have used the test framework by taking guidelines from angular2-seed project by mgechev. 
When i am debugging my unit test cases in phantomjs, instead of DebugRenderer, DOMRenderer is instanciated, because of which my test cases are failing. Please let me know what else i need to do.
